apps/meteor/app/reactions/server/setReaction.js

```diff
@@ -3,16 +3,40 @@ import { TAPi18n } from 'meteor/rocketchat:tap-i18n';
 import _ from 'underscore';
 import { EmojiCustom } from '@rocket.chat/models';
 
-import { Messages, Rooms } from '../../models/server';
+import { Messages, Rooms, Users } from '../../models/server';
 import { callbacks } from '../../../lib/callbacks';
 import { emoji } from '../../emoji/server';
 import { isTheLastMessage, msgStream } from '../../lib/server';
 import { canAccessRoom, hasPermission } from '../../authorization/server';
 import { api } from '../../../server/sdk/api';
 import { AppEvents, Apps } from '../../apps/server/orchestrator';
+import { settings } from '../../settings/server';
+
+const getNameByUsername = (username, message, reaction) => {
+	const names = new Map();
+	Users.findUsersByUsernames(message.reactions[reaction].usernames, {
+		fields: {
+			username: 1,
+			name: 1,
+		},
+	}).forEach((data) => {
+		names.set(data.username, data.name);
+	});
+	return names.get(username) || username;
+};
 
 const removeUserReaction = (message, reaction, username) => {
-	message.reactions[reaction].usernames.splice(message.reactions[reaction].usernames.indexOf(username), 1);
+	const index = message.reactions[reaction].usernames.indexOf(username);
+	message.reactions[reaction].usernames.splice(index, 1);
+	if (settings.get('UI_Use_Real_Name') && message.reactions[reaction].names) {
+		if (!message.reactions[reaction].names) {
+			message.reactions[reaction].names = message.reactions[reaction].usernames.map((username) =>
+				getNameByUsername(username, message, reaction),
+			);
+		} else {
+			message.reactions[reaction].names.splice(index, 1);
+		}
+	}
```

Comment:
You can avoid looping through twice with uncertain iterations by changing it from `reactions: {usernames: [], names: []}` to `reactions: [{username: '', name: ''}, ...]` and even sorting it in the backend.

Another problem with this is duplicate names. What if a user changes their name?

You're also in-place migrating? Should be a separate migration for this, not to be part of the code here.

The inner `if (true)` will never run.

Reply:
Yes it would make sense to use `[":reaction:"]: [{username: '', name: ''}, ...]` but I followed the existing type of [reaction](reaction) and the [`normalizeMessageForUser`](https://github.com/RocketChat/Rocket.Chat/blob/bf915717dbb1d0f4ba46df71f682fbb815ddcbe2/apps/meteor/app/utils/server/lib/normalizeMessagesForUser.ts#L74) function.

Is there a better way to achieve this?

Reply:
Update the type, and everywhere the structure is being used. 

I am also thinking we probably, instead of `[":reaction:"]: [{username: '', name: ''}, ...]`, use 
```js
[":reaction:"]: {
    [_id]: {
        username: '',
        name: ''
    },
    .
    .
    .
},
.
.
.
```

This way we won't have to iterate at all.

Reply:
I don't think we currently update the reaction usernames (consequently the names) once a user updates either. I'm thinking we could add a subscription for this, for each user, to keep track of the messages they've reacted to, and then update with a single query. Still would be expensive. 

@sampaiodiego wdyt?

Reply:
NB: Looked at how currently mentions are handled after those changes, and it directly uses the messages model.